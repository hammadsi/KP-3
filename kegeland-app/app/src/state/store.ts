import {combineReducers, configureStore} from '@reduxjs/toolkit';
import AsyncStorage from '@react-native-async-storage/async-storage';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import reduxFlipper from 'redux-flipper';

import appReducer from './ducks/app/app.reducer';
import bluetoothReducer from './ducks/bluetooth/bluetooth.reducer';
import authReducer from './ducks/auth/auth.reducer';
import questionsReducer from './ducks/questions/questions.reducer';
import sessionReducer from './ducks/session/session.reducer';

import rootSaga, {sagaMiddleware} from './rootSaga';

/**
 * Configuration for redux-persist.
 * Redux-persist is responsible for maintaining the state
 * between sessions.
 */
const persistConfig = {
  key: 'root',
  version: 1,
  storage: AsyncStorage,
  whitelist: ['app', 'auth', 'bluetooth'],
};

// Initiate the root reducer
export const rootReducer = combineReducers({
  app: appReducer,
  auth: authReducer,
  bluetooth: bluetoothReducer,
  questions: questionsReducer,
  session: sessionReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

// Check if development-environment is enabled
const isDev = !['production', 'test', 'integration'].includes(
  process.env.NODE_ENV || '',
);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        })
          .concat(logger)
          .concat(reduxFlipper())
          .concat(sagaMiddleware)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(sagaMiddleware),
  devTools: isDev,
});

// Start redux-saga
sagaMiddleware.run(rootSaga);

// Create persistor for persisting state
export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
