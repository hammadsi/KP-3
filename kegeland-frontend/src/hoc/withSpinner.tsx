/* eslint-disable react/display-name */
import { Flex, Spinner } from '@chakra-ui/react';
import { ComponentType, useEffect, useState } from 'react';

const withSpinner =
  <P extends object>(
    Component: ComponentType<P>,
    duration: number = 200,
  ): ComponentType<P> =>
  (props) => {
    const [spinner, setSpinner] = useState<boolean>(true);

    useEffect(() => {
      setTimeout(() => setSpinner(false), duration);
    }, []);
    return spinner ? (
      <Flex w="100%" h="100vh" justify="center" align="center">
        <Spinner size="xl" color="primary.600" />
      </Flex>
    ) : (
      <Component {...props} />
    );
  };

export default withSpinner;
