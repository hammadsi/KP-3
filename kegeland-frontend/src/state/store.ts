import { combineReducers, configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import authReducer from './ducks/auth/auth.reducer';
import patientsReducer from './ducks/patients/patients.reducer';
import questionnairesReducer from './ducks/questionnaires/questionnaires.reducer';
import { sensorReducer } from './ducks/sensors/sensors.reducer';
import sessionsReducer from './ducks/sessions/sessions.reducer';
import settingsReducer from './ducks/settings/settings.reducer';
import wheelchairPatientsReducer from './ducks/wheelchairPatients/wheelchairPatients.reducer';

const persistConfig = {
  key: 'root',
  version: 1,
  storage,
  whitelist: ['settings', 'auth'],
};

export const rootReducer = combineReducers({
  settings: settingsReducer,
  auth: authReducer,
  patients: patientsReducer,
  sensors: sensorReducer,
  sessions: sessionsReducer,
  questionnaires: questionnairesReducer,
  wheelchairPatients: wheelchairPatientsReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const isDev = !['production', 'test'].includes(process.env.NODE_ENV || '');

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(logger)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }),
  devTools: isDev,
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
