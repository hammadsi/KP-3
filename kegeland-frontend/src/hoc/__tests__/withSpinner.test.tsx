import { Text } from '@chakra-ui/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import React from 'react';

import withSpinner from '../withSpinner';
import { mockStore } from '../../state/mocks/store.mock';

jest.useFakeTimers();
jest.spyOn(global, 'setTimeout');

describe('test withSpinner', () => {
  const MockComponent: React.FC = () => {
    return <Text />;
  };

  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      value: () => {
        return {
          matches: false,
          addListener: () => {},
          removeListener: () => {},
        };
      },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly', () => {
    const Component = withSpinner(MockComponent, 200);
    const tree = renderer.create(
      <Provider store={mockStore()}>
        <MemoryRouter>
          <Component />
        </MemoryRouter>
      </Provider>,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render correctly with timer', () => {
    const Component = withSpinner(MockComponent, 200);
    const tree = renderer.create(
      <Provider store={mockStore()}>
        <MemoryRouter>
          <Component />
        </MemoryRouter>
      </Provider>,
    );
    jest.advanceTimersByTime(200);
    expect(tree.toJSON).toMatchSnapshot();
    expect(setTimeout).toHaveBeenCalledTimes(1);
  });

  it('should render correctly loding false', () => {
    const useStateMock: any = (init: boolean) => [init, jest.fn()];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    const Component = withSpinner(MockComponent, 200);
    const tree = renderer.create(
      <Provider store={mockStore()}>
        <MemoryRouter>
          <Component />
        </MemoryRouter>
      </Provider>,
    );
    jest.advanceTimersByTime(300);
    expect(tree.toJSON).toMatchSnapshot();
    expect(setTimeout).toHaveBeenCalledTimes(1);
  });
});
