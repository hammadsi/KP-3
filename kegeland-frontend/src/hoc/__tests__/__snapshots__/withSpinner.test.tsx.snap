// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test withSpinner should render correctly 1`] = `
<div
  className="css-lt6aye"
>
  <div
    className="chakra-spinner css-1vk4lrc"
  >
    <span
      className="css-8b45rq"
    >
      Loading...
    </span>
  </div>
</div>
`;

exports[`test withSpinner should render correctly loding false 1`] = `[Function]`;

exports[`test withSpinner should render correctly with timer 1`] = `[Function]`;
